docs/DeveloperGuide.md

```diff
@@ -1,38 +1,461 @@
+  * `CsvWriter` is responsible for handling the writing of data to the CSV data file.
+  * `CsvReader` is responsible for handling the reading of data from the CSV data file.
+
+### Common classes
+
+Classes used by multiple components are in the `seedu.stockpal.common` package.
+
+### Exception classes
+
+Exceptions classes used by multiple components are in the `seedu.stockpal.exceptions` package.
+
+--------------------------------------------------------------------------------------------------------------------
+
+## **Implementation**
+
+This section describes some noteworthy details on how certain features are implemented.
+
+### Add product feature
+
+The NewCommand class is responsible for adding a new product to the inventory in the StockPal application.
+
+#### Class diagram of NewCommand:
+<img src="images/AddCommandClass.png" alt=""/>
+
+
+**Attributes**
+* name: The name of the product.
+* quantity: The initial quantity of the product.
+* price: The price of the product.
+* description: The description of the product.
+
+**Methods**
+* `NewCommand`: Constructor for creating a new instance of the NewCommand class.
+* `execute`: Method to add the new product to the product list.
+* `createProduct`: Method to create a new product with a unique product ID.
+
+The following sequence diagram shows how an add operation works when the user inputs a valid `add` command:
+<img src="images/AddCommandSequence.png" alt="AddCommandSequence.png"/>
+
+The following sequence diagram details how `AddCommand#execute()` functions.
+<img src="images/AddCommandExecuteSequenceDiagram.png" alt="AddCommandExecuteSequenceDiagram.png"/>
+
+
+### Edit product feature
+#### Implementation
+The `edit` command is used to edit product details such as name, quantity, price and description.
+
+The edit product feature is facilitated by `EditCommand` which extends `Command`.
+
+The following sequence diagram summarizes what happens when a user inputs a valid `edit` command.
+<img src="images/EditCommandSequenceDiagram.png" alt="EditCommandSequenceDiagram.png"/>
+
+> INFO:
+> The lifeline for EditCommand should end at the destroy marker (X) but due to a limitation of PlantUML, 
+> the lifeline reaches the end of diagram.
+
+Validation of the user input is done in `Parser`, hence `EditCommand` assumes that all fields provided upon creation
+of a `EditCommand` object are properly formatted.
+
+Validation for other criteria are still carried out within `EditCommand`.
+1. Checking if at least 1 field (name, quantity, price or description) is provided.
+2. Checking if the product ID (PID) belongs to an existing product.
+
+Once all validation is completed, updating of product details is done by calling `ProductList#updateProduct()`.
+
+The following sequence diagram details how `EditCommand#execute()` functions.
+<img src="images/EditCommandExecuteSequenceDiagram.png" alt="EditCommandExecuteSequenceDiagram.png"/>
+
+**Aspect: Validating parameters and handling errors**
```

Comment:
Nice alternative suggestions, which give other developers the flexibility to handle error